File: bootstrap/windows/Installer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Net.Http;
using System.Security.Cryptography;
using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;
using System.Web.Script.Serialization;

namespace FixedByDesign {
  public sealed class Asset {
    public string name { get; set; }
    public string browser_download_url { get; set; }
    public long size { get; set; }
  }
  public sealed class Release {
    public string tag_name { get; set; }
    public bool draft { get; set; }
    public bool prerelease { get; set; }
    public Asset[] assets { get; set; }
  }
  public sealed class Payload {
    public string name { get; set; }
    public long size { get; set; }
    public string sha512 { get; set; }
  }
  public sealed class Manifest {
    public int schemaVersion { get; set; }
    public string version { get; set; }
    public Dictionary<string, Payload> payloads { get; set; }
  }
  public sealed class Installer : IDisposable {
    public const string Repository = "https://github.com/Fixed-By-Design/launcher";
    public const string Latest = "https://api.github.com/repos/Fixed-By-Design/launcher/releases/latest";
    const long MaximumPayload = 1024L * 1024 * 1024;
    readonly HttpClient client;
    public Installer(HttpMessageHandler handler = null) {
      ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12;
      client = new HttpClient(handler ?? new HttpClientHandler { AllowAutoRedirect = false });
      client.Timeout = TimeSpan.FromMinutes(20);
      client.DefaultRequestHeaders.UserAgent.ParseAdd("FixedByDesign-Setup/1");
      client.DefaultRequestHeaders.CacheControl = new System.Net.Http.Headers.CacheControlHeaderValue { NoCache = true };
    }
    public void Dispose() { client.Dispose(); }
    public static bool Trusted(Uri uri) {
      return uri.Scheme == "https" && uri.IsDefaultPort && uri.UserInfo == "" &&
        (uri.Host == "api.github.com" || uri.Host == "github.com" || uri.Host == "release-assets.githubusercontent.com" || uri.Host == "objects.githubusercontent.com");
    }
    async Task<HttpResponseMessage> Request(string url, CancellationToken cancellation) {
      var uri = new Uri(url);
      for (int redirect = 0; redirect < 6; redirect++) {
        if (!Trusted(uri)) throw new IOException("GitHub a renvoyé une adresse de téléchargement non autorisée.");
        var response = await client.GetAsync(uri, HttpCompletionOption.ResponseHeadersRead, cancellation);
        if ((int)response.StatusCode >= 300 && (int)response.StatusCode < 400) {
          var location = response.Headers.Location;
          response.Dispose();
          if (location == null) throw new IOException("Redirection GitHub invalide.");
          uri = location.IsAbsoluteUri ? location : new Uri(uri, location);
          continue;
        }
        if (response.IsSuccessStatusCode) return response;
        var status = response.StatusCode;
        response.Dispose();
        if (status == HttpStatusCode.NotFound) throw new IOException("Aucune version disponible. Réessaie plus tard.");
        if (status == HttpStatusCode.Forbidden || (int)status == 429) throw new IOException("GitHub limite les téléchargements. Patiente quelques minutes puis réessaie.");
        throw new IOException("GitHub est indisponible. Vérifie ta connexion puis réessaie.");
      }
      throw new IOException("Trop de redirections pendant le téléchargement.");
    }
    async Task<byte[]> Metadata(string url, CancellationToken cancellation) {
      using (var response = await Request(url, cancellation))
      using (var input = await response.Content.ReadAsStreamAsync())
      using (var output = new MemoryStream()) {
        var buffer = new byte[16384];
        int count;
        while ((count = await input.ReadAsync(buffer, 0, buffer.Length, cancellation)) > 0) {
          if (output.Length + count > 1024 * 1024) throw new IOException("Métadonnées GitHub trop volumineuses.");
          output.Write(buffer, 0, count);
        }
        return output.ToArray();
      }
    }
    public static Asset FindAsset(Release release, string name) {
      Asset found = null;
      foreach (var asset in release.assets ?? new Asset[0]) {
        if (asset.name != name) continue;
        if (found != null) throw new IOException("La release contient des fichiers en double.");
        found = asset;
      }
      var expected = Repository + "/releases/download/" + Uri.EscapeDataString(release.tag_name) + "/" + Uri.EscapeDataString(name);
      if (found == null || found.browser_download_url != expected) throw new IOException("La release est incomplète ou son adresse de téléchargement est invalide.");
      return found;
    }
    public static Payload Select(Release release, Manifest manifest) {
      if (release.draft || release.prerelease || !Regex.IsMatch(release.tag_name ?? "", @"\Av[0-9]+\.[0-9]+\.[0-9]+\z"))
        throw new IOException("Cette release ne correspond pas au canal de distribution.");
      if (manifest.schemaVersion != 1 || "v" + manifest.version != release.tag_name)
        throw new IOException("Le manifeste ne correspond pas à la release. Télécharge à nouveau l’installateur depuis GitHub.");
      Payload payload;
      if (manifest.payloads == null || !manifest.payloads.TryGetValue("windows-x64", out payload) || payload == null ||
          payload.name != "FBD-Launcher-" + manifest.version + "-windows-x64.exe" ||
          payload.size <= 0 || payload.size > MaximumPayload || !Regex.IsMatch(payload.sha512 ?? "", @"\A[a-f0-9]{128}\z"))
        throw new IOException("Le fichier Windows de cette version est invalide ou absent.");
      if (FindAsset(release, payload.name).size != payload.size) throw new IOException("La taille du fichier ne correspond pas au manifeste.");
      return payload;
    }
    public async Task<Tuple<Release, Payload>> Resolve(CancellationToken cancellation) {
      var json = new JavaScriptSerializer { MaxJsonLength = 1024 * 1024 };
      var release = json.Deserialize<Release>(System.Text.Encoding.UTF8.GetString(await Metadata(Latest, cancellation)));
      if (release == null || release.draft || release.prerelease || !Regex.IsMatch(release.tag_name ?? "", @"\Av[0-9]+\.[0-9]+\.[0-9]+\z"))
        throw new IOException("Aucune release compatible disponible.");
      var manifestAsset = FindAsset(release, "launcher-manifest.json");
      var manifest = json.Deserialize<Manifest>(System.Text.Encoding.UTF8.GetString(await Metadata(manifestAsset.browser_download_url, cancellation)));
      if (manifest == null) throw new IOException("Manifeste de téléchargement invalide.");
      return Tuple.Create(release, Select(release, manifest));
    }
    public async Task Download(Release release, Payload payload, string destination, IProgress<int> progress, CancellationToken cancellation) {
      var part = destination + ".part";
      try {
        using (var response = await Request(FindAsset(release, payload.name).browser_download_url, cancellation))
        using (var input = await response.Content.ReadAsStreamAsync())
        using (var output = new FileStream(part, FileMode.CreateNew, FileAccess.Write, FileShare.None, 65536, true))
        using (var hash = SHA512.Create()) {
          if (response.Content.Headers.ContentLength.HasValue && response.Content.Headers.ContentLength.Value != payload.size)
            throw new IOException("Taille de téléchargement inattendue.");
          var buffer = new byte[65536];
          long received = 0;
          int count;
          while ((count = await input.ReadAsync(buffer, 0, buffer.Length, cancellation)) > 0) {
            received += count;
            if (received > payload.size) throw new IOException("Le téléchargement dépasse la taille annoncée.");
            hash.TransformBlock(buffer, 0, count, buffer, 0);
            await output.WriteAsync(buffer, 0, count, cancellation);
            if (progress != null) progress.Report((int)(received * 100 / payload.size));
          }
          hash.TransformFinalBlock(new byte[0], 0, 0);
          var actual = BitConverter.ToString(hash.Hash).Replace("-", "").ToLowerInvariant();
          if (received != payload.size || actual != payload.sha512) throw new IOException("Le téléchargement est incomplet ou corrompu. Réessaie.");
          await output.FlushAsync(cancellation);
        }
        cancellation.ThrowIfCancellationRequested();
        File.Move(part, destination);
      } finally {
        if (File.Exists(part)) File.Delete(part);
      }
    }
    public static void MarkInternet(string path) {
      File.WriteAllText(path + ":Zone.Identifier", "[ZoneTransfer]\r\nZoneId=3\r\nHostUrl=" + Repository + "\r\n");
    }
  }
}

File: bootstrap/macos/Installer.swift
import Foundation
import CryptoKit
import AppKit
import Darwin

struct SetupError: LocalizedError {
    let message: String
    var errorDescription: String? { message }
}
struct ReleaseAsset: Decodable {
    let name: String
    let browser_download_url: String
    let size: Int64
}
struct Release: Decodable {
    let tag_name: String
    let draft: Bool
    let prerelease: Bool
    let assets: [ReleaseAsset]
}
struct Payload: Decodable {
    let name: String
    let size: Int64
    let sha512: String
}
struct Manifest: Decodable {
    let schemaVersion: Int
    let version: String
    let payloads: [String: Payload]
}

final class Transfer: NSObject, URLSessionDownloadDelegate {
    private let destination: URL
    private let maximum: Int64
    private let progress: (Double) -> Void
    private let lock = NSLock()
    private var cancelled = false
    private var task: URLSessionDownloadTask?
    private var continuation: CheckedContinuation<URL, Error>?
    private var failure: Error?
    private var completed = false
    init(destination: URL, maximum: Int64, progress: @escaping (Double) -> Void) {
        self.destination = destination
        self.maximum = maximum
        self.progress = progress
    }
    func fetch(_ url: URL) async throws -> URL {
        guard Installer.trusted(url) else { throw SetupError(message: "Adresse de téléchargement non autorisée.") }
        return try await withTaskCancellationHandler(operation: {
            try await withCheckedThrowingContinuation { continuation in
                let configuration = URLSessionConfiguration.ephemeral
                configuration.timeoutIntervalForRequest = 60
                configuration.timeoutIntervalForResource = 1200
                configuration.requestCachePolicy = .reloadIgnoringLocalCacheData
                let session = URLSession(configuration: configuration, delegate: self, delegateQueue: nil)
                var request = URLRequest(url: url)
                request.setValue("FixedByDesign-Setup/1", forHTTPHeaderField: "User-Agent")
                let task = session.downloadTask(with: request)
                lock.lock()
                self.continuation = continuation
                self.task = task
                let shouldCancel = cancelled
                lock.unlock()
                task.resume()
                if shouldCancel { task.cancel() }
            }
        }, onCancel: {
            self.lock.lock()
            self.cancelled = true
            self.task?.cancel()
            self.lock.unlock()
        })
    }
    func urlSession(_ session: URLSession, task: URLSessionTask, willPerformHTTPRedirection response: HTTPURLResponse, newRequest request: URLRequest, completionHandler: @escaping (URLRequest?) -> Void) {
        guard let url = request.url, Installer.trusted(url) else {
            failure = SetupError(message: "GitHub a renvoyé une redirection non autorisée.")
            completionHandler(nil)
            return
        }
        completionHandler(request)
    }
    func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didWriteData bytesWritten: Int64, totalBytesWritten: Int64, totalBytesExpectedToWrite: Int64) {
        guard totalBytesWritten <= maximum, totalBytesExpectedToWrite <= maximum else {
            failure = SetupError(message: "Le téléchargement dépasse la taille annoncée.")
            downloadTask.cancel()
            return
        }
        progress(min(1, Double(totalBytesWritten) / Double(maximum)))
    }
    func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didFinishDownloadingTo location: URL) {
        do {
            guard let response = downloadTask.response as? HTTPURLResponse, response.statusCode == 200 else {
                let status = (downloadTask.response as? HTTPURLResponse)?.statusCode
                throw SetupError(message: status == 404 ? "Aucune version disponible. Réessaie plus tard." : status == 403 || status == 429 ? "GitHub limite les téléchargements. Patiente puis réessaie." : "GitHub est indisponible. Vérifie ta connexion puis réessaie.")
            }
            let size = try location.resourceValues(forKeys: [.fileSizeKey]).fileSize ?? 0
            guard size <= maximum else { throw SetupError(message: "Téléchargement trop volumineux.") }
            try FileManager.default.moveItem(at: location, to: destination)
            completed = true
        } catch { failure = error }
    }
    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
        session.finishTasksAndInvalidate()
        if let error = failure ?? error { continuation?.resume(throwing: error) }
        else if completed { continuation?.resume(returning: destination) }
        else { continuation?.resume(throwing: SetupError(message: "Téléchargement incomplet.")) }
        continuation = nil
    }
}

enum Installer {
    static let repository = "https://github.com/Fixed-By-Design/launcher"
    static let latest = URL(string: "https://api.github.com/repos/Fixed-By-Design/launcher/releases/latest")!
    static let bundleID = "com.fixedbydesign.launcher"
    static func trusted(_ url: URL) -> Bool {
        url.scheme == "https" && (url.port == nil || url.port == 443) && url.user == nil && url.password == nil &&
        ["api.github.com", "github.com", "release-assets.githubusercontent.com", "objects.githubusercontent.com"].contains(url.host ?? "")
    }
    static func architecture() throws -> String {
        var arm: Int32 = 0
        var size = MemoryLayout<Int32>.size
        if sysctlbyname("hw.optional.arm64", &arm, &size, nil, 0) == 0 && arm == 1 { return "arm64" }
        var system = utsname()
        uname(&system)
        let machine = withUnsafePointer(to: &system.machine) {
            $0.withMemoryRebound(to: CChar.self, capacity: 1) { String(cString: $0) }
        }
        guard machine == "x86_64" else { throw SetupError(message: "Ce Mac n’est pas compatible avec les versions disponibles.") }
        return "x64"
    }
    static func asset(_ release: Release, _ name: String) throws -> ReleaseAsset {
        let matches = release.assets.filter { $0.name == name }
        let expected = repository + "/releases/download/" + release.tag_name + "/" + name
        guard matches.count == 1, matches[0].browser_download_url == expected else {
            throw SetupError(message: "La release est incomplète ou contient une adresse non autorisée.")
        }
        return matches[0]
    }
    static func select(_ release: Release, _ manifest: Manifest, _ architecture: String) throws -> Payload {
        guard !release.draft, !release.prerelease, release.tag_name.range(of: #"\Av[0-9]+\.[0-9]+\.[0-9]+\z"#, options: .regularExpression) != nil,
              manifest.schemaVersion == 1, "v" + manifest.version == release.tag_name, ["arm64", "x64"].contains(architecture) else {
            throw SetupError(message: "Le manifeste ne correspond pas au canal ou à ce Mac. Télécharge à nouveau l’installateur depuis GitHub.")
        }
        guard let payload = manifest.payloads["macos-" + architecture],
              payload.name == "FBD-Launcher-" + manifest.version + "-macos-" + architecture + ".dmg",
              payload.size > 0, payload.size <= 1024 * 1024 * 1024,
              payload.sha512.range(of: #"\A[a-f0-9]{128}\z"#, options: .regularExpression) != nil,
              try asset(release, payload.name).size == payload.size else {
            throw SetupError(message: "Le fichier macOS est invalide ou absent.")
        }
        return payload
    }
    static func metadata<T: Decodable>(_ url: URL, directory: URL) async throws -> T {
        let path = directory.appendingPathComponent(UUID().uuidString + ".json")
        _ = try await Transfer(destination: path, maximum: 1024 * 1024, progress: { _ in }).fetch(url)
        return try JSONDecoder().decode(T.self, from: Data(contentsOf: path))
    }
    static func resolve(directory: URL) async throws -> (Release, Payload, String) {
        let release: Release = try await metadata(latest, directory: directory)
        guard !release.draft, !release.prerelease, release.tag_name.range(of: #"\Av[0-9]+\.[0-9]+\.[0-9]+\z"#, options: .regularExpression) != nil else {
            throw SetupError(message: "Aucune release compatible disponible.")
        }
        let manifestAsset = try asset(release, "launcher-manifest.json")
        guard let url = URL(string: manifestAsset.browser_download_url) else { throw SetupError(message: "Adresse de manifeste invalide.") }
        let manifest: Manifest = try await metadata(url, directory: directory)
        return (release, try select(release, manifest, architecture()), manifest.version)
    }
    static func verify(_ file: URL, payload: Payload) throws {
        let input = try FileHandle(forReadingFrom: file)
        defer { try? input.close() }
        var hash = SHA512()
        var size: Int64 = 0
        while let data = try input.read(upToCount: 65536), !data.isEmpty {
            try Task.checkCancellation()
            size += Int64(data.count)
            guard size <= payload.size else { throw SetupError(message: "Le téléchargement dépasse la taille annoncée.") }
            hash.update(data: data)
        }
        let actual = hash.finalize().map { String(format: "%02x", $0) }.joined()
        guard size == payload.size, actual == payload.sha512 else { throw SetupError(message: "Le téléchargement est incomplet ou corrompu. Réessaie.") }
    }
    static func command(_ executable: String, _ arguments: [String]) throws {
        let process = Process()
        process.executableURL = URL(fileURLWithPath: executable)
        process.arguments = arguments
        process.standardOutput = FileHandle.nullDevice
        process.standardError = FileHandle.nullDevice
        try process.run()
        process.waitUntilExit()
        guard process.terminationStatus == 0 else { throw SetupError(message: "macOS n’a pas pu terminer la vérification ou l’installation (\(URL(fileURLWithPath: executable).lastPathComponent), code \(process.terminationStatus)).") }
    }
    static func markInternet(_ path: URL) throws {
        let value = "0081;\(String(Int(Date().timeIntervalSince1970), radix: 16));FixedByDesignSetup;\(UUID().uuidString)"
        let result = value.withCString { setxattr(path.path, "com.apple.quarantine", $0, strlen($0), 0, 0) }
        guard result == 0 else { throw SetupError(message: "Impossible de conserver la provenance Internet du téléchargement.") }
    }
    static func replace(_ staged: URL, destination: URL) throws {
        let exists = FileManager.default.fileExists(atPath: destination.path)
        let result = exists ? renamex_np(staged.path, destination.path, UInt32(RENAME_SWAP)) : rename(staged.path, destination.path)
        guard result == 0 else { throw SetupError(message: "Impossible de remplacer l’application. La version existante est conservée.") }
    }
    static func install(_ image: URL, version: String, directory: URL, applications: URL? = nil) throws -> URL {
        let files = FileManager.default
        let applications = applications ?? files.homeDirectoryForCurrentUser.appendingPathComponent("Applications", isDirectory: true)
        try files.createDirectory(at: applications, withIntermediateDirectories: true)
        let destination = applications.appendingPathComponent("Fixed By Design.app", isDirectory: true)
        if files.fileExists(atPath: destination.path) {
            guard !((try destination.resourceValues(forKeys: [.isSymbolicLinkKey])).isSymbolicLink ?? false),
                  Bundle(url: destination)?.bundleIdentifier == bundleID else {
                throw SetupError(message: "Un autre fichier occupe déjà ~/Applications/Fixed By Design.app. Il n’a pas été modifié.")
            }
        }
        guard NSRunningApplication.runningApplications(withBundleIdentifier: bundleID).isEmpty else {
            throw SetupError(message: "Ferme Fixed By Design et Minecraft, puis réessaie.")
        }
        let mount = directory.appendingPathComponent("image", isDirectory: true)
        try files.createDirectory(at: mount, withIntermediateDirectories: true)
        try command("/usr/bin/hdiutil", ["attach", "-readonly", "-nobrowse", "-noautoopen", "-mountpoint", mount.path, image.path])
        var mounted = true
        let staging = applications.appendingPathComponent(".fbd-setup-" + UUID().uuidString, isDirectory: true)
        do {
            let source = mount.appendingPathComponent("Fixed By Design.app", isDirectory: true)
            guard Bundle(url: source)?.bundleIdentifier == bundleID,
                  Bundle(url: source)?.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String == version else {
                throw SetupError(message: "L’application de cette image ne correspond pas à la version annoncée.")
            }
            if let minimum = Bundle(url: source)?.object(forInfoDictionaryKey: "LSMinimumSystemVersion") as? String {
                let components = minimum.split(separator: ".").compactMap { Int($0) }
                guard components.count >= 2, components.count <= 3,
                      ProcessInfo.processInfo.isOperatingSystemAtLeast(OperatingSystemVersion(majorVersion: components[0], minorVersion: components[1], patchVersion: components.count == 3 ? components[2] : 0)) else {
                    throw SetupError(message: "Cette version du launcher nécessite macOS \(minimum) ou ultérieur. Ton installation existante est conservée.")
                }
            }
            try command("/usr/bin/codesign", ["--verify", "--deep", "--strict", source.path])
            try files.createDirectory(at: staging, withIntermediateDirectories: false)
            let staged = staging.appendingPathComponent("Fixed By Design.app", isDirectory: true)
            try files.copyItem(at: source, to: staged)
            try command("/usr/bin/codesign", ["--verify", "--deep", "--strict", staged.path])
            try markInternet(staged)
            try Task.checkCancellation()
            try command("/usr/bin/hdiutil", ["detach", mount.path])
            mounted = false
            try replace(staged, destination: destination)
            try files.removeItem(at: staging)
            return destination
        } catch {
            var cleanup = ""
            if mounted {
                do { try command("/usr/bin/hdiutil", ["detach", mount.path]) }
                catch { cleanup += " L’image reste montée : \(mount.path)." }
            }
            if files.fileExists(atPath: staging.path) {
                do { try files.removeItem(at: staging) }
                catch { cleanup += " Le dossier temporaire reste présent : \(staging.path)." }
            }
            throw SetupError(message: error.localizedDescription + cleanup)
        }
    }
}
